src/redisraft.c

```diff
@@ -509,6 +509,34 @@ static void handleMigrateCommand(RedisRaftCtx *rr, RedisModuleCtx *ctx, RaftRedi
     raft_entry_release(entry);
 }
 
+static bool getAskingState(RedisRaftCtx *rr, RedisModuleCtx *ctx)
+{
+    ClientState *clientState = ClientStateGet(rr, ctx);
+    return clientState->asking;
+}
+
+static void setAskingState(RedisRaftCtx *rr, RedisModuleCtx *ctx, bool val)
+{
+    ClientState *clientState = ClientStateGet(rr, ctx);
+    clientState->asking = val;
+}
+
+static void handleAsking(RedisRaftCtx *rr, RedisModuleCtx *ctx)
+{
+    Node *leader_proxy = NULL;
+
+    Node **ret = rr->config->follower_proxy ? &leader_proxy : NULL;
+    if (checkLeader(rr, ctx, ret) == RR_ERROR) {
+        if (!rr->config->follower_proxy) {
+            return;
+        }
+    }
```

Comment:
This check won't work, please see the usage in `handleRedisCommand()`.

Other than that, probably we don't even need to handle this. Is there a reason we are handling this? We can just set the client "asking" whether we are leader or not.

Reply:
so, the Q is, if we are not leader, how should we handle an ASKING.

do we 

a) accept it and handle leader later on a "real" command
b) reject it and reply moved back to migrating shardgroup
c) reject it and reply moved to our leader
d) reject it and reply ask to our leader

you're arguing for a. 

c. I'm not sure makes sense because the asking won't transfer with it.  (clients expect that a moved response means they are being redirected to primary owner, i.e. the migrating shard) 
b. is impossible because we dont know who it is (i.e. asking doesn't give us a slot)
d. would be the "best" (i.e. you're asking us, but you want our leader).  But unsure we can do it right (as dont know slot).

now if we do it later, we have 2 options

1) replyMoved back to migrating shard
2) since we know that we are an asking, replyAsk as appropriate (with correct slot info) to our leader.  I'm a bit skeptical on this.

but I'm ok with your suggestion, just wanted to enumerate everything.

Reply:
another approach would be to try treat an ASK to a follower as an inherent follower proxy mode?  i.e. we will proxy it to right node, as we can't "nicely" redirect you to it.

Reply:
As I said, if this scenario is super rare, we shouldn't optimize for it. What happens if we don't do this? If this doesn't cause any visible improvement, I'd say we shouldn't write code for it.

Reply:
I'm not sure if its super rare, it be behavior I'd expect to see in real life usage.  I actually think as long as we have a proxy mode, it makes sense to leverage it for asking.

Reply:
i.e. with the recent understanding that asking resets the state after every execution, every asking will be an independent command returned from an ask response.

with proxy we avoid a lot of ping ponging and just handle it naturally.  the reason to not support proxy normally is to teach the client where to go so it we can avoid the man in the middle.  there wont be ping poinging.  However, for ask/asking thats irrelvant.

With that said, can discuss it tomorrow